### PR TITLE
FIX ShadowPEFT last-token pooling with left padding

### LINKED ISSUE #3620: [Bug] ShadowPEFT sequence classification pools the wrong token with left padding
### System Info

- PEFT `main`: `9c16ee66cd4c58bd9cdf2d8b4e06c1cf8e8f8efe`
- Affected code: the ShadowPEFT sequence-classification pooling path introduced in #3354.

### Who can help?

@BenjaminBossan

### Reproduction

ShadowPEFT documents `_pool_last_token` as selecting the last non-padding representation, but it derives the index as `attention_mask.sum(dim=1) - 1`:

```python
import torch

from peft.tuners.shadow.model import _pool_last_token

hidden = torch.tensor(
    [
        [[10.0], [20.0], [30.0], [40.0]],
        [[10.0], [20.0], [30.0], [40.0]],
    ]
)
attention_mask = torch.tensor(
    [
        [0, 0, 1, 1],
        [0, 1, 1, 1],
    ]
)

print(_pool_last_token(hidden, attention_mask).squeeze(-1))
```

The current index expression selects positions 1 and 2, producing `[20, 30]`. Both are incorrect for these left-padded rows. The last non-padding position is 3 in each row, so the expected result is `[40, 40]`.

The same count-based implementation appears twice:

- the helper used by `ShadowModel.shadow_auxiliary_loss`, so the auxiliary sequence-classification loss trains against a padding position;
- `DetachedShadowModel._pool_last_token`, so `unload_shadow()` inference returns logits from the wrong position.

Transformers sequence-classification models intentionally select the rightmost non-padding position to support both left and right padding:

https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_layers.py

Tokenizer configuration also exposes `padding_side` as a supported choice:

https://huggingface.co/docs/transformers/main_classes/tokenizer

I searched PEFT issues, pull requests, and history for ShadowPEFT padding, classification pooling, left padding, and `_pool_last_token`. I found no overlapping report or implementation.

### Expected behavior

ShadowPEFT sequence classification should select the rightmost position whose attention-mask value is nonzero, matching Transformers for both left- and right-padded batches.

The proposed scope is one shared pooling helper used by both training and detached inference, with focused regressions for left and right padding. Causal-language-model and cache behavior would remain unchanged.

Would you like me to prepare that focused PR? I will wait for explicit approval before changing code.

AI assistance was used during the source audit and duplicate search. I reviewed the complete affected path and independently checked the index calculation against the documented Transformers behavior.

### PR DESCRIPTION
**Pull request by aswanth-07**
Fixes #3620.

## Summary

ShadowPEFT sequence classification derived the pooled index from the number of valid tokens. That selects the correct token for right padding, but it can select a padding position when valid tokens are left padded.

This change selects the rightmost non-padding position instead. A shared functional helper now owns the pooling rule used by both the auxiliary training loss and `DetachedShadowModel`, which keeps training and standalone inference consistent. The existing behavior for right padding, an all-padding row, and a missing attention mask remains unchanged.

## Coordination

- Issue discussion: https://github.com/huggingface/peft/issues/3620
- Maintainer approval: https://github.com/huggingface/peft/issues/3620#issuecomment-5476438684

## Tests

- `ruff check src/peft/tuners/shadow/layers.py src/peft/tuners/shadow/model.py tests/test_shadow.py` passed with Ruff 0.16.4.
- `ruff format --check src/peft/tuners/shadow/layers.py src/peft/tuners/shadow/model.py tests/test_shadow.py` passed with Ruff 0.16.4.
- `git diff --check` passed.
- `pytest tests/test_shadow.py -k classification_pooling_respects_padding_side -v` was not run locally because this checkout has no compatible Torch test environment. The focused regression is included for CI validation.

## AI assistance

AI assistance disclosure: AI-assisted development tools were used during investigation and implementation. The reported tests were run against the final diff.

The contributor reviewed every changed line. The focused runtime regression remains for CI validation.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@SeanLee97 It would be great if you could check.

File: src/peft/tuners/shadow/layers.py [hunk 1]
```diff
@@ -38,6 +38,16 @@ def _sequence_classification_loss(logits: torch.Tensor, labels: torch.Tensor) ->
     return torch.nn.functional.binary_cross_entropy_with_logits(logits, labels)
 
 
+def _pool_last_token(hidden: torch.Tensor, attention_mask: Optional[torch.Tensor]) -> torch.Tensor:
+    """Pool the last non-padding token representation for sequence classification."""
+    if attention_mask is None:
+        return hidden[:, -1, :]
+    token_positions = torch.arange(attention_mask.shape[-1], device=attention_mask.device)
+    last_non_padding_token = (attention_mask.ne(0) * token_positions).argmax(dim=-1)
```
Comment (BenjaminBossan) [inline src/peft/tuners/shadow/layers.py]: > `last_non_padding_token`

`last_non_padding_token_idx` or `last_non_padding_token_indices` would be more fitting, right?

File: src/peft/tuners/shadow/layers.py [hunk 2]
```diff
@@ -38,6 +38,16 @@ def _sequence_classification_loss(logits: torch.Tensor, labels: torch.Tensor) ->
     return torch.nn.functional.binary_cross_entropy_with_logits(logits, labels)
 
 
+def _pool_last_token(hidden: torch.Tensor, attention_mask: Optional[torch.Tensor]) -> torch.Tensor:
+    """Pool the last non-padding token representation for sequence classification."""
```
Comment (BenjaminBossan) [inline src/peft/tuners/shadow/layers.py]: I know the docstring was already like this before, but let's clarify that we want to return the hidden state of the last non-padding for _both left and right padding_ cases. "Pool" is a bit ambiguous.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the PR, the fix LGTM and corresponds roughly to what we have in Transformers:

https://github.com/huggingface/transformers/blob/03a4d7b91e06cccdd0b2c68eb5c282a048f5dfb0/src/transformers/modeling_layers.py#L163-L164

Just some nits on naming and docstring from my side.

File: src/peft/tuners/shadow/layers.py (hunk 1 above)
Comment (aswanth-07) [reply src/peft/tuners/shadow/layers.py]: Updated this to `last_non_padding_token_indices` in 8e9d828e to reflect that the tensor contains one index per batch item.

File: src/peft/tuners/shadow/layers.py (hunk 2 above)
Comment (aswanth-07) [reply src/peft/tuners/shadow/layers.py]: Clarified the docstring in 8e9d828e. The helper now states that it returns the hidden state of the last non-padding token for both left- and right-padded sequences.

**Comment by aswanth-07** [comment]
@BenjaminBossan Thanks for the review. I addressed both nits in 8e9d828e. Ruff check and format check pass for the three changed files. The focused runtime regression remains for CI validation because this local checkout does not have a compatible Torch test environment.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for the latest fixes, LGTM.
